### PR TITLE
feat: Add invariant hooks, refactor to handle multiple suites (SC-481)

### PR DESCRIPTION
**Pull request by lucas-manuel**

### CODE REVIEW

File: test/invariant/handlers/TransferHandler.sol [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    constructor(
+        PSM3      psm_,
+        MockERC20 asset0,
+        MockERC20 asset1,
+        MockERC20 asset2
+    ) HandlerBase(psm_, asset0, asset1, asset2) {}
+
+    function transfer(uint256 assetSeed, string memory senderSeed, uint256 amount) external {
+        MockERC20 asset = _getAsset(assetSeed);
+        address   sender = makeAddr(senderSeed);
```
Comment (hexonaut) [inline test/invariant/handlers/TransferHandler.sol:23]: Alignment

File: test/invariant/handlers/HandlerBase.sol [hunk 2]
```diff
@@ -15,6 +15,8 @@ contract HandlerBase is CommonBase, StdCheatsSafe, StdUtils {
 
     MockERC20[3] public assets;
 
+    uint256 public constant TRILLION = 1e12;
```
Comment (hexonaut) [inline test/invariant/handlers/HandlerBase.sol:18]: It's a bit odd to have this as a constant variable in the base. This is not being reused in the sub-classes except for 1 and it's not an arbitrary value like `NUM_TOKENS` or something. I would just use 1e12 in `LpHandler` directly, and one million like you are in the `TransferHandler`.

File: test/invariant/handlers/TransferHandler.sol [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+        // Bounding to 1 million here because 1 trillion introduces unrealistic conditions with
+        // large rounding errors. Would rather keep tolerances smaller with a lower upper bound
+        // on transfer amounts.
+        amount = _bound(amount, 1, 1_000_000 * 10 ** asset.decimals());
```
Comment (hexonaut) [inline test/invariant/handlers/TransferHandler.sol:28]: So this is meant to represent accidental token donations? If so then ya 1m is reasonable.

**Review by hexonaut — CHANGES_REQUESTED** [review]
Mostly minor stuff. Really good assertions here. I ran with 100 runs at 5000 depth with no issue.

File: test/invariant/handlers/TransferHandler.sol (hunk 3 above)
Comment (lucas-manuel) [reply test/invariant/handlers/TransferHandler.sol:28]: Yeah figured this was a reasonable upper bound for that

Comment (lucas-manuel) [reply test/invariant/handlers/TransferHandler.sol:28]: Also could be for malicious intent

File: test/invariant/handlers/HandlerBase.sol (hunk 2 above)
Comment (lucas-manuel) [reply test/invariant/handlers/HandlerBase.sol:18]: Updated

File: test/invariant/handlers/TransferHandler.sol (hunk 1 above)
Comment (lucas-manuel) [reply test/invariant/handlers/TransferHandler.sol]: Fixed

File: test/invariant/handlers/HandlerBase.sol (hunk 2 above)
Comment (hexonaut) [reply test/invariant/handlers/HandlerBase.sol:18]: Still need to delete the constant.

**Review by hexonaut — APPROVED** [review]

**Pull request merged**